sequencers/single/single.go

```diff
@@ -0,0 +1,402 @@
+package single
+
+import (
+	"bytes"
+	"context"
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"cosmossdk.io/log"
+	ds "github.com/ipfs/go-datastore"
+
+	coreda "github.com/rollkit/rollkit/core/da"
+	coresequencer "github.com/rollkit/rollkit/core/sequencer"
+	dac "github.com/rollkit/rollkit/da"
+	pb "github.com/rollkit/rollkit/types/pb/rollkit"
+)
+
+// ErrInvalidRollupId is returned when the rollup id is invalid
+var (
+	ErrInvalidRollupId = errors.New("invalid rollup id")
+
+	initialBackoff = 100 * time.Millisecond
+
+	maxSubmitAttempts = 30
+	defaultMempoolTTL = 25
+)
+
+var _ coresequencer.Sequencer = &Sequencer{}
+
+// Sequencer implements go-sequencing interface
+type Sequencer struct {
+	logger log.Logger
+
+	rollupId  []byte
+	dalc      coreda.Client
+	batchTime time.Duration
+
+	maxBlobSize            *atomic.Uint64
+	lastSubmittedBatchHash atomic.Value
+	seenBatches            sync.Map
+
+	// submitted batches
+	sbq *BatchQueue
+	// pending batches
+	bq *BatchQueue
+
+	metrics *Metrics
+}
+
+// NewSequencer creates a new Centralized Sequencer
+func NewSequencer(
+	ctx context.Context,
+	logger log.Logger,
+	db ds.Batching,
+	da coreda.DA,
+	daNamespace,
+	rollupId []byte,
+	batchTime time.Duration,
+	metrics *Metrics,
+) (*Sequencer, error) {
+	dalc := dac.NewDAClient(da, -1, -1, daNamespace, nil, logger)
+	mBlobSize, err := dalc.MaxBlobSize(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	maxBlobSize := &atomic.Uint64{}
+	maxBlobSize.Store(mBlobSize)
+
+	s := &Sequencer{
+		logger:      logger,
+		dalc:        dalc,
+		batchTime:   batchTime,
+		maxBlobSize: maxBlobSize,
+		rollupId:    rollupId,
+		// TODO: this can cause an overhead with IO
+		bq:          NewBatchQueue(db, "pending"),
+		sbq:         NewBatchQueue(db, "submitted"),
+		seenBatches: sync.Map{},
+		metrics:     metrics,
+	}
+
+	err = s.bq.Load(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to load batch queue from DB: %w", err)
+	}
+
+	err = s.sbq.Load(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to load submitted batch queue from DB: %w", err)
+	}
+
+	go s.batchSubmissionLoop(ctx)
+	return s, nil
+}
+
+// SubmitRollupBatchTxs implements sequencing.Sequencer.
+func (c *Sequencer) SubmitRollupBatchTxs(ctx context.Context, req coresequencer.SubmitRollupBatchTxsRequest) (*coresequencer.SubmitRollupBatchTxsResponse, error) {
+	if !c.isValid(req.RollupId) {
+		return nil, ErrInvalidRollupId
+	}
+
+	// create a batch from the set of transactions and store it in the queue
+	batch := coresequencer.Batch{Transactions: req.Batch.Transactions}
+	err := c.bq.AddBatch(ctx, batch)
+	if err != nil {
+		return nil, fmt.Errorf("failed to add batch: %w", err)
+	}
+
+	lastSubmittedBatchHash := c.lastSubmittedBatchHash.Load()
```

Comment:
`lastSubmittedBatchHas` is not very useful. I recommend getting rid of it. Note that, in rollkit you are not using it currently. Also, in the future, this needs to be changed to batch data (which is the <da height + da commitment>), which can only be obtained after publishing the batch to DA, where you would update the sbq batch with this information so that GetNextBatch can get the submitted batch along with this information to be inserted into the Rollkit header while proposing.